File: src/js/extra/content_processor.js
// Advanced Content Processing for AI Crawlers
// Implements Strategy and Template patterns for content extraction and response building

// Content Extraction Strategies
class ContentExtractionStrategy {
    extractContent(originalResponse, headers) {
        throw new Error('Method must be implemented by subclass');
    }
    
    getStrategyName() {
        throw new Error('Method must be implemented by subclass');
    }
}

class ResponseBuildingStrategy {
    buildResponse(content, template) {
        throw new Error('Method must be implemented by subclass');
    }
    
    getStrategyName() {
        throw new Error('Method must be implemented by subclass');
    }
}

// Data structures for extracted content
class ExtractedContent {
    constructor({
        title = '',
        headings = [],
        mainContent = '',
        keywords = [],
        links = [],
        images = [],
        metadata = {},
        summary = '',
        language = 'en',
        contentType = 'text/html'
    } = {}) {
        this.title = title;
        this.headings = headings;
        this.mainContent = mainContent;
        this.keywords = keywords;
        this.links = links;
        this.images = images;
        this.metadata = metadata;
        this.summary = summary;
        this.language = language;
        this.contentType = contentType;
    }
}

class Link {
    constructor(url, text, title = '', isExternal = false, relationship = '') {
        this.url = url;
        this.text = text;
        this.title = title;
        this.isExternal = isExternal;
        this.relationship = relationship;
    }
}

class Image {
    constructor(url, alt = '', title = '', width = 0, height = 0) {
        this.url = url;
        this.alt = alt;
        this.title = title;
        this.width = width;
        this.height = height;
    }
}

// Response Template interface
class ResponseTemplate {
    getTemplate() {
        throw new Error('Method must be implemented by subclass');
    }
    
    setVariables(variables) {
        this.variables = variables;
    }
    
    render(content) {
        throw new Error('Method must be implemented by subclass');
    }
}

// Content Extraction Strategy Implementations

class HTMLContentExtractor extends ContentExtractionStrategy {
    extractContent(originalResponse, headers) {
        // TODO: Implement HTML parsing and content extraction
        // - Parse HTML using DOM parser or cheerio-like library
        // - Extract title, headings (h1-h6), main content
        // - Extract meta tags and structured data
        // - Identify and extract links and images
        // - Generate content summary
        // - Extract keywords and key phrases
        return new ExtractedContent();
    }
    
    getStrategyName() {
        return 'HTML Content Extractor';
    }
}

class JSONContentExtractor extends ContentExtractionStrategy {
    extractContent(originalResponse, headers) {
        // TODO: Implement JSON parsing and content extraction
        // - Parse JSON structure
        // - Extract relevant data fields
        // - Convert to standardized ExtractedContent format
        return new ExtractedContent();
    }
    
    getStrategyName() {
        return 'JSON Content Extractor';
    }
}

class XMLContentExtractor extends ContentExtractionStrategy {
    extractContent(originalResponse, headers) {
        // TODO: Implement XML parsing and content extraction
        // - Parse XML structure
        // - Extract relevant elements and attributes
        // - Convert to standardized ExtractedContent format
        return new ExtractedContent();
    }
    
    getStrategyName() {
        return 'XML Content Extractor';
    }
}

// Response Building Strategy Implementations

class SummaryBuilder extends ResponseBuildingStrategy {
    buildResponse(content, template) {
        // TODO: Implement summary response building
        // - Create concise summaries of the content
        // - Focus on key information and main points
        // - Optimize for AI understanding and processing
        return '';
    }
    
    getStrategyName() {
        return 'Summary Builder';
    }
}

class StructuredDataBuilder extends ResponseBuildingStrategy {
    buildResponse(content, template) {
        // TODO: Implement structured data response building
        // - Add JSON-LD structured data
        // - Enhance with schema.org markup
        // - Optimize for rich snippets and knowledge graphs
        return '';
    }
    
    getStrategyName() {
        return 'Structured Data Builder';
    }
}

// Response Template Implementations

class BaseTemplate extends ResponseTemplate {
    constructor(templateString = '') {
        super();
        this.templateString = templateString;
        this.variables = {};
    }
    
    getTemplate() {
        return this.templateString;
    }
    
    render(content) {
        // TODO: Implement base template rendering
        // - Replace template variables with actual content
        // - Handle template logic and conditionals
        return '';
    }
}

class AIOptimizedTemplate extends BaseTemplate {
    render(content) {
        // TODO: Implement AI-optimized template rendering
        // - Structure content for optimal AI consumption
        // - Add semantic markup and context
        // - Include relevant metadata and relationships
        return '';
    }
}

class MinimalTemplate extends BaseTemplate {
    render(content) {
        // TODO: Implement minimal template rendering
        // - Clean, minimal HTML output
        // - Focus on content without decorative elements
        // - Optimize for fast processing
        return '';
    }
}

// Main Content Processor Class

class ContentProcessor {
    constructor(extractionStrategy, buildingStrategy, template) {
        this.extractionStrategy = extractionStrategy;
        this.buildingStrategy = buildingStrategy;
        this.template = template;
    }
    
    async processContent(originalResponse, headers) {
        // TODO: Implement the main content processing pipeline
        // 1. Extract content using the configured extraction strategy
        // 2. Build response using the configured building strategy and template
        // 3. Return the optimized response
        return '';
    }
    
    setExtractionStrategy(strategy) {
        this.extractionStrategy = strategy;
    }
    
    setBuildingStrategy(strategy) {
        this.buildingStrategy = strategy;
    }
    
    setTemplate(template) {
        this.template = template;
    }
}

// Factory Functions

function getExtractionStrategy(contentType) {
    // TODO: Implement strategy selection based on content type
    // - Return HTMLContentExtractor for HTML content
    // - Return JSONContentExtractor for JSON content
    // - Return XMLContentExtractor for XML content
    if (contentType && contentType.includes('application/json')) {
        return new JSONContentExtractor();
    } else if (contentType && (contentType.includes('application/xml') || contentType.includes('text/xml'))) {
        return new XMLContentExtractor();
    } else {
        return new HTMLContentExtractor();
    }
}

function getBuildingStrategy(strategyType) {
    // TODO: Implement strategy selection based on requirements
    switch (strategyType) {
        case 'summary':
            return new SummaryBuilder();
        case 'structured':
            return new StructuredDataBuilder();
        default:
            return new SummaryBuilder();
    }
}

function getTemplate(templateType) {
    // TODO: Implement template selection based on type
    switch (templateType) {
        case 'ai-optimized':
            return new AIOptimizedTemplate();
        case 'minimal':
            return new MinimalTemplate();
        default:
            return new BaseTemplate();
    }
}

// Export all classes and functions
export {
    ContentExtractionStrategy,
    ResponseBuildingStrategy,
    ExtractedContent,
    Link,
    Image,
    ResponseTemplate,
    HTMLContentExtractor,
    JSONContentExtractor,
    XMLContentExtractor,
    SummaryBuilder,
    StructuredDataBuilder,
    BaseTemplate,
    AIOptimizedTemplate,
    MinimalTemplate,
    ContentProcessor,
    getExtractionStrategy,
    getBuildingStrategy,
    getTemplate
};

File: src/python/extra/content_processor.py
# Advanced Content Processing for AI Crawlers
# Implements Strategy and Template patterns for content extraction and response building

from abc import ABC, abstractmethod
from typing import Dict, List, Optional, Any, Union
from dataclasses import dataclass, field
from enum import Enum

# Data structures for extracted content

@dataclass
class Link:
    """Represents a link found in the content"""
    url: str
    text: str
    title: str = ''
    is_external: bool = False
    relationship: str = ''

@dataclass
class Image:
    """Represents an image found in the content"""
    url: str
    alt: str = ''
    title: str = ''
    width: int = 0
    height: int = 0

@dataclass
class ExtractedContent:
    """Holds the parsed content from the original response"""
    title: str = ''
    headings: List[str] = field(default_factory=list)
    main_content: str = ''
    keywords: List[str] = field(default_factory=list)
    links: List[Link] = field(default_factory=list)
    images: List[Image] = field(default_factory=list)
    metadata: Dict[str, str] = field(default_factory=dict)
    summary: str = ''
    language: str = 'en'
    content_type: str = 'text/html'

# Strategy Interfaces

class ContentExtractionStrategy(ABC):
    """Abstract base class for content extraction strategies"""
    
    @abstractmethod
    def extract_content(self, original_response: str, headers: Dict[str, str]) -> ExtractedContent:
        """Extract content from the original response"""
        pass
    
    @abstractmethod
    def get_strategy_name(self) -> str:
        """Get the name of this strategy"""
        pass

class ResponseBuildingStrategy(ABC):
    """Abstract base class for response building strategies"""
    
    @abstractmethod
    def build_response(self, content: ExtractedContent, template: 'ResponseTemplate') -> str:
        """Build a response using the extracted content and template"""
        pass
    
    @abstractmethod
    def get_strategy_name(self) -> str:
        """Get the name of this strategy"""
        pass

# Template Interface

class ResponseTemplate(ABC):
    """Abstract base class for response templates"""
    
    def __init__(self):
        self.variables: Dict[str, str] = {}
    
    @abstractmethod
    def get_template(self) -> str:
        """Get the template string"""
        pass
    
    def set_variables(self, variables: Dict[str, str]) -> None:
        """Set template variables"""
        self.variables = variables
    
    @abstractmethod
    def render(self, content: ExtractedContent) -> str:
        """Render the template with the given content"""
        pass

# Content Extraction Strategy Implementations

class HTMLContentExtractor(ContentExtractionStrategy):
    """Extracts content from HTML responses"""
    
    def extract_content(self, original_response: str, headers: Dict[str, str]) -> ExtractedContent:
        """
        TODO: Implement HTML parsing and content extraction
        - Parse HTML using BeautifulSoup or similar library
        - Extract title, headings (h1-h6), main content
        - Extract meta tags and structured data
        - Identify and extract links and images
        - Generate content summary
        - Extract keywords and key phrases
        """
        return ExtractedContent()
    
    def get_strategy_name(self) -> str:
        return "HTML Content Extractor"

class JSONContentExtractor(ContentExtractionStrategy):
    """Extracts content from JSON responses"""
    
    def extract_content(self, original_response: str, headers: Dict[str, str]) -> ExtractedContent:
        """
        TODO: Implement JSON parsing and content extraction
        - Parse JSON structure
        - Extract relevant data fields
        - Convert to standardized ExtractedContent format
        """
        return ExtractedContent()
    
    def get_strategy_name(self) -> str:
        return "JSON Content Extractor"

class XMLContentExtractor(ContentExtractionStrategy):
    """Extracts content from XML responses"""
    
    def extract_content(self, original_response: str, headers: Dict[str, str]) -> ExtractedContent:
        """
        TODO: Implement XML parsing and content extraction
        - Parse XML structure using lxml or xml.etree
        - Extract relevant elements and attributes
        - Convert to standardized ExtractedContent format
        """
        return ExtractedContent()
    
    def get_strategy_name(self) -> str:
        return "XML Content Extractor"

# Response Building Strategy Implementations

class SummaryBuilder(ResponseBuildingStrategy):
    """Builds concise summary responses"""
    
    def build_response(self, content: ExtractedContent, template: ResponseTemplate) -> str:
        """
        TODO: Implement summary response building
        - Create concise summaries of the content
        - Focus on key information and main points
        - Optimize for AI understanding and processing
        """
        return ""
    
    def get_strategy_name(self) -> str:
        return "Summary Builder"

class StructuredDataBuilder(ResponseBuildingStrategy):
    """Builds responses with enhanced structured data"""
    
    def build_response(self, content: ExtractedContent, template: ResponseTemplate) -> str:
        """
        TODO: Implement structured data response building
        - Add JSON-LD structured data
        - Enhance with schema.org markup
        - Optimize for rich snippets and knowledge graphs
        """
        return ""
    
    def get_strategy_name(self) -> str:
        return "Structured Data Builder"

# Response Template Implementations

class BaseTemplate(ResponseTemplate):
    """Base template implementation"""
    
    def __init__(self, template_string: str = ''):
        super().__init__()
        self.template_string = template_string
    
    def get_template(self) -> str:
        return self.template_string
    
    def render(self, content: ExtractedContent) -> str:
        """
        TODO: Implement base template rendering
        - Replace template variables with actual content
        - Handle template logic and conditionals
        """
        return ""

class AIOptimizedTemplate(BaseTemplate):
    """Template optimized for AI crawler consumption"""
    
    def render(self, content: ExtractedContent) -> str:
        """
        TODO: Implement AI-optimized template rendering
        - Structure content for optimal AI consumption
        - Add semantic markup and context
        - Include relevant metadata and relationships
        """
        return ""

class MinimalTemplate(BaseTemplate):
    """Minimal, clean template implementation"""
    
    def render(self, content: ExtractedContent) -> str:
        """
        TODO: Implement minimal template rendering
        - Clean, minimal HTML output
        - Focus on content without decorative elements
        - Optimize for fast processing
        """
        return ""

# Main Content Processor Class

class ContentProcessor:
    """Main class that orchestrates content processing using Strategy and Template patterns"""
    
    def __init__(self, 
                 extraction_strategy: ContentExtractionStrategy,
                 building_strategy: ResponseBuildingStrategy,
                 template: ResponseTemplate):
        self.extraction_strategy = extraction_strategy
        self.building_strategy = building_strategy
        self.template = template
    
    async def process_content(self, original_response: str, headers: Dict[str, str]) -> str:
        """
        Process the original response and return an AI-optimized response
        
        TODO: Implement the main content processing pipeline
        1. Extract content using the configured extraction strategy
        2. Build response using the configured building strategy and template
        3. Return the optimized response
        """
        return ""
    
    def set_extraction_strategy(self, strategy: ContentExtractionStrategy) -> None:
        """Change the extraction strategy at runtime"""
        self.extraction_strategy = strategy
    
    def set_building_strategy(self, strategy: ResponseBuildingStrategy) -> None:
        """Change the building strategy at runtime"""
        self.building_strategy = strategy
    
    def set_template(self, template: ResponseTemplate) -> None:
        """Change the template at runtime"""
        self.template = template

# Factory Functions

def get_extraction_strategy(content_type: str) -> ContentExtractionStrategy:
    """
    Get the appropriate extraction strategy based on content type
    
    TODO: Implement strategy selection based on content type
    - Return HTMLContentExtractor for HTML content
    - Return JSONContentExtractor for JSON content
    - Return XMLContentExtractor for XML content
    """
    if content_type and 'application/json' in content_type:
        return JSONContentExtractor()
    elif content_type and ('application/xml' in content_type or 'text/xml' in content_type):
        return XMLContentExtractor()
    else:
        return HTMLContentExtractor()

def get_building_strategy(strategy_type: str) -> ResponseBuildingStrategy:
    """
    Get the appropriate building strategy based on requirements
    
    TODO: Implement strategy selection based on requirements
    """
    strategy_map = {
        'summary': SummaryBuilder,
        'structured': StructuredDataBuilder
    }
    
    strategy_class = strategy_map.get(strategy_type, SummaryBuilder)
    return strategy_class()

def get_template(template_type: str) -> ResponseTemplate:
    """
    Get the appropriate template based on template type
    
    TODO: Implement template selection based on type
    """
    template_map = {
        'ai-optimized': AIOptimizedTemplate,
        'minimal': MinimalTemplate,
        'base': BaseTemplate
    }
    
    template_class = template_map.get(template_type, BaseTemplate)
    return template_class()

# Convenience function to create a content processor with default settings
def create_default_content_processor() -> ContentProcessor:
    """Create a content processor with default settings"""
    return ContentProcessor(
        extraction_strategy=HTMLContentExtractor(),
        building_strategy=SummaryBuilder(),
        template=AIOptimizedTemplate()
    )
